apps/cowswap-frontend/src/modules/combinedBalances/hooks/useTokensBalancesCombined.ts

```diff
@@ -0,0 +1,60 @@
+import { useMemo } from 'react'
+
+import { BalancesState, useTokensBalances } from '@cowprotocol/balances-and-allowances'
+import { useWalletInfo } from '@cowprotocol/wallet'
+
+import { BigNumber } from 'ethers'
+
+import { usePreHookBalanceDiff } from 'modules/hooksStore/hooks/useBalancesDiff'
+import { useIsHooksTradeType } from 'modules/trade'
+
+export function useTokensBalancesCombined() {
+  const { account } = useWalletInfo()
+  const preHooksBalancesDiff = usePreHookBalanceDiff()
+  const tokenBalances = useTokensBalances()
+  const isHooksTradeType = useIsHooksTradeType()
+
+  return useMemo(() => {
+    if (!account || !isHooksTradeType) return tokenBalances
+    const accountBalancesDiff = preHooksBalancesDiff[account.toLowerCase()] || {}
+    return applyBalanceDiffs(tokenBalances, accountBalancesDiff)
+  }, [account, preHooksBalancesDiff, tokenBalances, isHooksTradeType])
+}
+
+function applyBalanceDiffs(currentState: BalancesState, diffs: Record<string, string>): BalancesState {
+  const normalizedValues: Record<string, BigNumber> = {}
+
+  // Create normalized versions of inputs
+  const normalizedCurrentValues: Record<string, BigNumber | undefined> = {}
+  const normalizedDiffs: Record<string, string> = {}
+
+  // Normalize current state values
+  for (const address of Object.keys(currentState.values)) {
+    normalizedCurrentValues[address] = currentState.values[address]
+  }
+
+  // Normalize diffs
+  for (const address of Object.keys(diffs)) {
+    normalizedDiffs[address] = diffs[address]
+  }
```

Comment:
nitpick: it fees a bit more complex than needed. Would something in this line work?

```js
const balanceEntries = concurrentState.values.entries().concat(diff.entries)
return balanceEntries.reduce(([address, balance]) => {
   ...
)
````